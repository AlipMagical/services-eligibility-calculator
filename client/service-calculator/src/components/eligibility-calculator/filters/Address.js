/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Card, Col, Row } from 'react-bootstrap';
import React, { useState } from 'react';

import { Chip } from '../Chip';
import { ExternalLink } from '../../../common/utils';
import FilterModal from './FilterModal';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { useTranslation } from 'react-i18next';

function AddressCard({ children, onClick, selected, materialIcon }) {
  return (
    <Col xs={12} className="card-col">
      <Card
        onClick={onClick}
        as="button"
        className={selected ? 'selected' : ''}
        aria-pressed={selected ? 'true' : 'false'}
      >
        <Card.Body>
          <i className="material-icons card-icon align-top" aria-hidden="true">
            {materialIcon}
          </i>
          <span className="card-text align-middle">{children}</span>
        </Card.Body>
      </Card>
    </Col>
  );
}

function sendAnalytics(resident) {
  ReactGA.event({
    category: 'Filter',
    action: 'Resident',
    label: resident.toString()
  });
}

function AddressFilterModal({ resident, setResident, setShow }) {
  const [localState, setLocalState] = useState(resident);
  const { t } = useTranslation();

  return (
    <FilterModal
      className="address"
      show
      onApply={() => {
        setResident(localState);
        setShow(false);
        localState !== undefined && sendAnalytics(localState);
      }}
      onHide={() => setShow(false)}
      onClear={() => {
        setShow(false);
        setResident(undefined);
      }}
      clearDisabled={localState == null}
      title={t('filterAddress.modalTitle')}
    >
      <Row>
        <Col xs={12} className="mt-2 mb-3">
          <span className="input-label">
            {t('filterAddress.modalDescription')}
          </span>
        </Col>
      </Row>
      <Row className="card-row">
        <AddressCard
          onClick={() => setLocalState(true)}
          selected={localState === true}
          materialIcon="check"
        >
          {t('filterAddress.resident')}
        </AddressCard>
        <AddressCard
          onClick={() => setLocalState(false)}
          selected={localState === false}
          materialIcon="close"
        >
          {t('filterAddress.notResident')}
        </AddressCard>
      </Row>
      <Row>
        <Col xs={12} className="mt-3 text-center">
          <ExternalLink
            className="no-text-decoration"
            url={t('filterAddress.unsureLink')}
            action="ResidencyDetermination"
          >
            <span className="unsure-link">
              {t('filterAddress.unsureLinkText')}{' '}
              <i className="material-icons align-top" aria-hidden="true">
                open_in_new
              </i>
            </span>
          </ExternalLink>
        </Col>
      </Row>
    </FilterModal>
  );
}

export default function Address({ resident, setResident }) {
  const [show, setShow] = useState(false);
  const { t } = useTranslation();

  return (
    <>
      <Chip
        analytics={{ action: 'Address', label: 'Add address' }}
        active={resident != null}
        onClick={() => setShow(true)}
      >
        {resident == null && t('filterAddress.buttonAdd')}
        {resident === true && t('filterAddress.buttonCity')}
        {resident === false && t('filterAddress.buttonNotCity')}
      </Chip>
      {show && <AddressFilterModal {...{ resident, setResident, setShow }} />}
    </>
  );
}

Address.propTypes = {
  resident: PropTypes.bool,
  setResident: PropTypes.func.isRequired
};
