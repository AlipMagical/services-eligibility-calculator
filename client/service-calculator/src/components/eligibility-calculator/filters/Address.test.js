/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { axe, toHaveNoViolations } from 'jest-axe';
import { fireEvent, render, waitForElement } from '@testing-library/react';

import Address from './Address';
import React from 'react';
import { act } from 'react-dom/test-utils';

expect.extend(toHaveNoViolations);

it('renders button without selected resident state', async () =>
  act(async () => {
    const { container } = render(
      <Address resident={undefined} setResident={() => {}} />
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal without selected resident state', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <Address resident={undefined} setResident={() => {}} />
    );

    fireEvent.click(getByText('Add address'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));

it('renders button with selected resident state', async () =>
  act(async () => {
    const { container } = render(
      <Address resident={true} setResident={() => {}} />
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal with selected resident state', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <Address resident={true} setResident={() => {}} />
    );

    fireEvent.click(getByText('In City'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));

it('renders button with selected not resident state', async () =>
  act(async () => {
    const { container } = render(
      <Address resident={false} setResident={() => {}} />
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal with selected not resident state', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <Address resident={false} setResident={() => {}} />
    );

    fireEvent.click(getByText('Not in City'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));
