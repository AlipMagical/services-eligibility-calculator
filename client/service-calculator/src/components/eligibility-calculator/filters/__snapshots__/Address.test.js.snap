// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button with selected not resident state 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis active"
  >
    Not in City
  </button>
</div>
`;

exports[`renders button with selected resident state 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis active"
  >
    In City
  </button>
</div>
`;

exports[`renders button without selected resident state 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis "
  >
    Add address
  </button>
</div>
`;

exports[`renders modal with selected not resident state 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Address
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="mt-2 mb-3 col-12"
        >
          <span
            class="input-label"
          >
            A city address is required for most services
          </span>
        </div>
      </div>
      <div
        class="card-row row"
      >
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                check
              </i>
              <span
                class="card-text align-middle"
              >
                Resident of City
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="true"
            class="selected card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                close
              </i>
              <span
                class="card-text align-middle"
              >
                Not a resident of City
              </span>
            </div>
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="mt-3 text-center col-12"
        >
          <a
            class="no-text-decoration"
            href="https://maps.google.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="unsure-link"
            >
              I'm not sure, help me check
               
              <i
                aria-hidden="true"
                class="material-icons align-top"
              >
                open_in_new
              </i>
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;

exports[`renders modal with selected resident state 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Address
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="mt-2 mb-3 col-12"
        >
          <span
            class="input-label"
          >
            A city address is required for most services
          </span>
        </div>
      </div>
      <div
        class="card-row row"
      >
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="true"
            class="selected card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                check
              </i>
              <span
                class="card-text align-middle"
              >
                Resident of City
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                close
              </i>
              <span
                class="card-text align-middle"
              >
                Not a resident of City
              </span>
            </div>
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="mt-3 text-center col-12"
        >
          <a
            class="no-text-decoration"
            href="https://maps.google.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="unsure-link"
            >
              I'm not sure, help me check
               
              <i
                aria-hidden="true"
                class="material-icons align-top"
              >
                open_in_new
              </i>
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;

exports[`renders modal without selected resident state 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        Address
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="row"
      >
        <div
          class="mt-2 mb-3 col-12"
        >
          <span
            class="input-label"
          >
            A city address is required for most services
          </span>
        </div>
      </div>
      <div
        class="card-row row"
      >
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                check
              </i>
              <span
                class="card-text align-middle"
              >
                Resident of City
              </span>
            </div>
          </button>
        </div>
        <div
          class="card-col col-12"
        >
          <button
            aria-pressed="false"
            class="card"
          >
            <div
              class="card-body"
            >
              <i
                aria-hidden="true"
                class="material-icons card-icon align-top"
              >
                close
              </i>
              <span
                class="card-text align-middle"
              >
                Not a resident of City
              </span>
            </div>
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="mt-3 text-center col-12"
        >
          <a
            class="no-text-decoration"
            href="https://maps.google.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="unsure-link"
            >
              I'm not sure, help me check
               
              <i
                aria-hidden="true"
                class="material-icons align-top"
              >
                open_in_new
              </i>
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="disabled btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
